## Travis tests have failed

Hey @gaocegege,
Please read the following log in order to understand the failure reason.
It'll be awesome if you fix what's wrong and commit the changes.


### Ruby: undefined


<details>
  <summary>
    <strong>
     make validate
    </strong>
  </summary>

```
xelatex -no-pdf -halt-on-error thesis
This is XeTeX, Version 3.14159265-2.6-0.99999 (TeX Live 2018) (preloaded format=xelatex)
 restricted \write18 enabled.
entering extended mode
(./thesis.tex
LaTeX2e <2018-04-01> patch level 4
(./sjtuthesis.cls
Document Class: sjtuthesis 2018/01/09 v0.10 Shanghai Jiao Tong University Templ
ate
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/ctexbook.cls
(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/expl3.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/expl3-code.tex)
(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/l3xdvipdfmx.def))
Document Class: ctexbook 2018/05/02 v2.4.14 Chinese adapter for class book (CTE
X)
(/usr/local/texlive/2018/texmf-dist/tex/latex/l3packages/xparse/xparse.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/l3packages/l3keys2e/l3keys2e.sty)
 (/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/ctexhook.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/ctexpatch.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/fix-cm.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1enc.def))
(/usr/local/texlive/2018/texmf-dist/tex/latex/ms/everysel.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/config/ctexopts.cfg)
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/book.cls
Document Class: book 2014/09/29 v1.4h Standard LaTeX document class
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/bk12.clo))
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/engine/ctex-engine-xetex.def
(/usr/local/texlive/2018/texmf-dist/tex/xelatex/xecjk/xeCJK.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/l3packages/xtemplate/xtemplate.st
y) (/usr/local/texlive/2018/texmf-dist/tex/latex/fontspec/fontspec.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/fontspec/fontspec-xetex.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/fontenc.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/tuenc.def))
(/usr/local/texlive/2018/texmf-dist/tex/latex/fontspec/fontspec.cfg)))
(/usr/local/texlive/2018/texmf-dist/tex/xelatex/xecjk/xeCJK.cfg))
(/usr/local/texlive/2018/texmf-dist/tex/xelatex/xecjk/xeCJKfntef.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/ulem/ulem.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/environ/environ.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/trimspaces/trimspaces.sty))
(/usr/local/texlive/2018/texmf-dist/tex/latex/cjk/texinput/CJKfntef.sty)))
(/usr/local/texlive/2018/texmf-dist/tex/latex/zhnumber/zhnumber.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/zhnumber/zhnumber-utf8.cfg))
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/scheme/ctex-scheme-chinese-b
ook.def
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/config/ctex-name-utf8.cfg))
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/ctex-cs4size.clo)
(/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/fontset/ctex-fontset-fandol.
def
*************************************************
* fontspec warning: "script-not-exist"
* 
* Font 'FandolSong-Regular' does not contain script 'CJK'.
*************************************************
)) (/usr/local/texlive/2018/texmf-dist/tex/latex/ctex/config/ctex.cfg)
(/usr/local/texlive/2018/texmf-dist/tex/latex/geometry/geometry.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/keyval.sty)
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifpdf.sty)
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifvtex.sty)
(/usr/local/texlive/2018/texmf-dist/tex/generic/ifxetex/ifxetex.sty))
(/usr/local/texlive/2018/texmf-dist/tex/latex/fancyhdr/fancyhdr.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/lastpage/lastpage.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/mathtools/mathtools.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/calc.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/mathtools/mhsetup.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsmath.sty
For additional information on amsmath, use the `?' option.
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amstext.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsgen.sty))
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsbsy.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsopn.sty)))
(/usr/local/texlive/2018/texmf-dist/tex/latex/amscls/amsthm.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsfonts/amsfonts.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/amsfonts/amssymb.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/bm.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/newtx/newtxtext.sty
`newtxtext' v1.531, 2018/03/27 Text macros taking advantage of TeX-Gyre Termes 
fonts (msharpe)
(/usr/local/texlive/2018/texmf-dist/tex/latex/fontaxes/fontaxes.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/xkeyval/xkeyval.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/xkeyval/xkeyval.tex
(/usr/local/texlive/2018/texmf-dist/tex/generic/xkeyval/xkvutils.tex)))
(/usr/local/texlive/2018/texmf-dist/tex/latex/etoolbox/etoolbox.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/fontenc.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/t1enc.def)
(/usr/local/texlive/2018/texmf-dist/tex/latex/lm/t1lmr.fd))
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/textcomp.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1enc.def)))
(/usr/local/texlive/2018/texmf-dist/tex/latex/newtx/newtxmath.sty
`newtxmath' v1.55, 2018/05/09 Math macros based originally on txfonts (msharpe)
(/usr/local/texlive/2018/texmf-dist/tex/latex/etextools/etextools.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/etex-pkg/etex.sty

Package etex Warning: Extended allocation already in use.
(etex)                etex.sty code will not be used.
(etex)                To force etex package to load, add
(etex)                \RequirePackage{etex}
(etex)                at the start of the document.

) (/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/letltxmacro.sty))
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifluatex.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/centernot.sty)
(/usr/local/texlive/2018/texmf-dist/tex/generic/kastrup/binhex.tex))
(/usr/local/texlive/2018/texmf-dist/tex/latex/sourcecodepro/sourcecodepro.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/newtx/t1ntxtlf.fd))
(/usr/local/texlive/2018/texmf-dist/tex/latex/was/upgreek.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/wasysym/wasysym.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/anyfontsize/anyfontsize.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/metalogo/metalogo.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/graphicx.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/graphics.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/trig.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)
(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-def/xetex.def))))
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/doc.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/multicol.sty))
(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/array.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/threeparttable/threeparttable.sty
) (/usr/local/texlive/2018/texmf-dist/tex/latex/tools/dcolumn.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/multirow/multirow.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/booktabs/booktabs.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/caption.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/caption3.sty))
(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/bicaption.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/subcaption.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/biblatex/biblatex.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/pdftexcmds.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/infwarerr.sty)
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ltxcmds.sty))
(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/kvoptions.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/kvsetkeys.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/etexcmds.sty)))
(/usr/local/texlive/2018/texmf-dist/tex/latex/logreq/logreq.sty
(/usr/local/texlive/2018/texmf-dist/tex/latex/logreq/logreq.def))
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ifthen.sty)
(/usr/local/texlive/2018/texmf-dist/tex/latex/url/url.sty)
(/usr/local/texlive/2018/texmf-dist/tex/generic/xstring/xstring.sty
(/usr/local/texlive/2018/texmf-dist/tex/generic/xstring/xstring.tex))
(/usr/local/texlive/2018/texmf-dist/tex/latex/biblatex/blx-dm.def)
! Parameters must be numbered consecutively.
<to be read again> 
                   1
l.6132 \def\do#1
                {%
No pages of output.
Transcript written on thesis.log.
make: *** [validate] Error 1
```

</details>

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -1,0 +1,3 @@
+version: v0.10.{build}
+os: Visual Studio 2017
+clone_folder: c:\projects\SJTUThesis

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -1,3 +1,28 @@
 version: v0.10.{build}
 os: Visual Studio 2017
 clone_folder: c:\projects\SJTUThesis
+
+install:
+  - curl -L http://mirror.ctan.org/systems/texlive/tlnet/install-tl.zip > c:\projects\install-tl.zip
+  - cd c:\projects && unzip c:\projects\install-tl.zip -d c:\projects\tmp
+  # - xcopy c:\projects\tmp\* c:\projects\install-tl
+  - unzip c:\projects\install-tl.zip
+  # Use echo to send a CRLF to the program to stop hanging on pause command.
+  # See https://stackoverflow.com/questions/33374713/how-to-skip-the-procedure-needed-to-press-any-key-to-exit-the-exe-in-batch-file
+  - echo . | c:\projects\install-tl-20180513\install-tl-windows.bat --profile c:\projects\SJTUThesis\.ci\tl.profile
+
+before_build:
+  - cd c:\projects\SJTUThesis
+  - sed -i 's/,\s*fontset=\?\w*//' thesis.tex
+  - sed -i '/setmainfont/d' sjtuthesis.cls
+  - sed -i 's/documentclass\[/documentclass[fontset=fandol, /' thesis.tex
+
+build_script:
+  # - set VERSION=$(c:\projects\install-tl-20180512\install-tl-windows.bat --version | grep 'version' | grep -o '[0-9]\{4\}')
+  # - PLATFORM=$(c:\projects\install-tl-20180512\install-tl-windows.bat --print-platform)
+  - set TEXBIN="c:\texlive\2018\bin\win32"
+  - set PATH=%PATH%;%TEXBIN%
+  - tlmgr install latexmk ulem fandol
+  - cd c:\projects\SJTUThesis
+  - xelatex -no-pdf -halt-on-error thesis
+  - biber --debug thesis

--- a/.ci/tl.profile
+++ b/.ci/tl.profile
@@ -1,0 +1,14 @@
+selected_scheme scheme-basic
+collection-basic 1
+collection-bibtexextra 1
+collection-fontsextra 1
+collection-fontsrecommended 1
+collection-langchinese 1
+collection-latex 1
+collection-latexextra 1
+collection-latexrecommended 1
+collection-mathscience 1
+collection-xetex 1
+tlpdbopt_autobackup 0
+tlpdbopt_install_docfiles 0
+tlpdbopt_install_srcfiles 0

--- a/.travis.yml
+++ b/.travis.yml
@@ -9,22 +9,7 @@ install:
   - mkdir -p $INSTALL
   - curl -sSL $REMOTE/install-tl-unx.tar.gz | tar -xzv -C $INSTALL --strip-components=1
 
-  - echo "selected_scheme scheme-basic" >> $INSTALL/tl.profile
-  - echo "collection-basic 1" >> $INSTALL/tl.profile
-  - echo "collection-bibtexextra 1" >> $INSTALL/tl.profile
-  - echo "collection-fontsextra 1" >> $INSTALL/tl.profile
-  - echo "collection-fontsrecommended 1" >> $INSTALL/tl.profile
-  - echo "collection-langchinese 1" >> $INSTALL/tl.profile
-  - echo "collection-latex 1" >> $INSTALL/tl.profile
-  - echo "collection-latexextra 1" >> $INSTALL/tl.profile
-  - echo "collection-latexrecommended 1" >> $INSTALL/tl.profile
-  - echo "collection-mathscience 1" >> $INSTALL/tl.profile
-  - echo "collection-xetex 1" >> $INSTALL/tl.profile
-  - echo "tlpdbopt_autobackup 0" >> $INSTALL/tl.profile
-  - echo "tlpdbopt_install_docfiles 0" >> $INSTALL/tl.profile
-  - echo "tlpdbopt_install_srcfiles 0" >> $INSTALL/tl.profile
-
-  - sudo $INSTALL/install-tl -profile $INSTALL/tl.profile
+  - sudo $INSTALL/install-tl -profile .ci/tl.profile
 
   - VERSION=$($INSTALL/install-tl --version | grep 'version' | grep -o '[0-9]\{4\}')
   - PLATFORM=$($INSTALL/install-tl --print-platform)

